src/macros.jl

```diff
@@ -502,77 +411,55 @@ par2
 ```
 """
 macro finite_parameter(args...)
-    # process the inputs
-    pos_args, kwargs, container_type, base_name = _extract_kwargs(args)
-
     # make an error function
-    _error(str...) = _macro_error(:finite_parameter, (args...,), 
-                                  __source__, str...)
+    error_fn = JuMPC.build_error_fn(:finite_parameter, args, __source__)
+
+    # process the inputs
+    pos_args, kwargs = JuMPC.parse_macro_arguments(
+        error_fn, 
+        args, 
+        num_positional_args = 2,
+        valid_kwargs = [:container, :base_name]
+    )
 
     # process the positional arguments
-    isempty(pos_args) && _error("No model was given.")
-    model = popfirst!(pos_args)
-    esc_model = esc(model)
-    if length(pos_args) == 1
-        expr = popfirst!(pos_args)
-    else
-        _error("Incorrect number of arguments. Must be of form ",
-               "@finite_parameter(model, name_expr == value).")
-    end
+    model_sym = popfirst!(pos_args)
+    model = esc(model_sym)
+    expr = popfirst!(pos_args)
     if isexpr(expr, :call)
         if expr.args[1] !== :(==)
-            _error("Unrecognized operator. Must be of form ",
+            error_fn("Unrecognized operator. Must be of form ",
                    "@finite_parameter(model, name_expr == value).")
         end
         param = expr.args[2]
         value = _esc_non_constant(expr.args[3])
-        is_anon = isexpr(param, (:vect, :vcat))
     else
-        param = gensym()
+        param = nothing
         value = _esc_non_constant(expr)
-        is_anon = true
     end
 
-    # process the name
-    name = _get_name(param)
-    if isnothing(base_name)
-        base_name = is_anon ? "" : string(name)
-    end
-    if !isa(name, Symbol) && !is_anon
-        _error("Expression $name should not be used as a parameter name. Use " *
-               "the \"anonymous\" syntax $name = @finite_parameter(model, " *
-               "...) instead.")
+    # make sure param is something reasonable
+    if !(param isa Union{Nothing, Symbol, Expr})
+        error_fn("Expected `$param` to be a parameter name.")
```

Comment:
So we could throw this error from `parse_ref_sets`. But then it isn't obvious how we would give a good error for `to be a parameter name` replacing `parameter` for something meaningful in each case.

Reply:
I thought about this, I suppose we could since in all cases the name is actually used to create a Julia variable the stores a reference or a container of references.

Reply:
I think it's a method error for now, so this is always something that could be improved in future.